client/client.go

```diff
@@ -24,14 +43,32 @@ func NewClient(cnf Config) (*Client, error) {
 		cnf.HttpClient.SetBaseURL(cnf.HostURL)
 	}
 
-	if cnf.ServiceToken == "" {
-		return nil, fmt.Errorf("you must set the service token for the client before making calls")
+	if cnf.ServiceToken == "" && cnf.ClientId == "" && cnf.ClientSecret == "" {
+		return nil, fmt.Errorf("you must set the service token, or a client secret and client ID for the client before making calls")
 	}
 
-	if cnf.ServiceToken != "" {
-		cnf.HttpClient.SetAuthToken(cnf.ServiceToken)
+	var authToken string
+
+	if cnf.ClientId != "" && cnf.ClientSecret != "" {
+		token, err := Client{cnf}.UniversalMachineIdentityAuth()
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to authenticate with universal machine identity [err=%s]", err)
+		}
+
+		authToken = token
+		cnf.AuthStrategy = AuthStrategy.UNIVERSAL_MACHINE_IDENTITY
+	}
+	if cnf.ServiceToken != "" && authToken == "" {
```

Comment:
I think we should add a check to make sure they aren't trying to set multiple auth methods at once. What do you think?

Reply:
I agree. I refactored the client auth code to be more readable and gave it a better validation flow